### PR TITLE
FV version of multiphase, multicomponent PorousFlow materials

### PR DESCRIPTION
**Pull request by cpgr**
This is the next part of making PorousFlow work with FV, and hopefully the hardest!

### CODE REVIEW

**Comment by cpgr** [comment]
I think that with these changes I'll be able to add the example for #24590.

**Comment by GiudGiud** [comment]
I should be able to take a first look this weekend.

**Comment by cpgr** [comment]
You can leave it until next week!

File: modules/porous_flow/src/materials/PorousFlowFluidStateBaseMaterial.C
Comment (GiudGiud) [inline modules/porous_flow/src/materials/PorousFlowFluidStateBaseMaterial.C:103]: there is no AD old material properties though, the derivatives wouldnt match the current value of the solution

File: modules/porous_flow/src/materials/PorousFlowFluidStateBaseMaterial.C
Comment (GiudGiud) [inline modules/porous_flow/src/materials/PorousFlowFluidStateBaseMaterial.C:151]: since you're going to fill those, you dont need this assign, you could have resize everywhere instead of assign and save the "setting to 0" operation

Comment (GiudGiud) [reply modules/porous_flow/src/materials/PorousFlowFluidStateBaseMaterial.C:151]: kind of a micro-optimization though. Porbably not needed

**Review by GiudGiud — COMMENTED** [review]
quick look, skipped two large files

I think the handling of AD is mostly similar to what we did elsewhere.
I ll need to look more at the FV stuff

File: modules/porous_flow/src/materials/PorousFlowFluidStateBaseMaterial.C
Comment (GiudGiud) [reply modules/porous_flow/src/materials/PorousFlowFluidStateBaseMaterial.C:103]: ok I saw they're not declared, just a nullptr

File: modules/porous_flow/include/materials/PorousFlowFluidStateBaseMaterial.h
Comment (cpgr) [reply modules/porous_flow/include/materials/PorousFlowFluidStateBaseMaterial.h:66]: Yeah, they are needed for the non-AD version, but are set to nullptr for the AD version

File: modules/porous_flow/include/materials/PorousFlowFluidStateBaseMaterial.h
Comment (cpgr) [reply modules/porous_flow/include/materials/PorousFlowFluidStateBaseMaterial.h:89]: Do you mean define `usingFooMembers` and a shorter `usingADFooMembers`? Then is it just
```
if (is_ad)
   usingADFooMembers
else
  usingFooMembers
```

File: modules/porous_flow/include/userobjects/PorousFlowBrineCO2.h
Comment (cpgr) [reply modules/porous_flow/include/userobjects/PorousFlowBrineCO2.h:456]: I thought that ADReal and DualReal are the same now, and that ADReal was preferred?

File: modules/porous_flow/src/userobjects/PorousFlowWaterVapor.C
Comment (cpgr) [reply modules/porous_flow/src/userobjects/PorousFlowWaterVapor.C:43]: Yeah, I'm not sure what is best here. There is only one fluid component (water) that can exist in either the liquid or vapour phase. Maybe I should just use a single `_fluid_component` to make it clear? I could just use 0 but this makes it similar to the other, multiple component classes.

File: modules/porous_flow/src/userobjects/PorousFlowWaterVapor.C
Comment (cpgr) [reply modules/porous_flow/src/userobjects/PorousFlowWaterVapor.C:214]: Yeah, not the vapour fraction (saturation).

File: modules/porous_flow/src/materials/PorousFlowFluidStateBaseMaterial.C
Comment (cpgr) [reply modules/porous_flow/src/materials/PorousFlowFluidStateBaseMaterial.C:54]: Hehehe - I've thought for years that I should delete the 'phase' from the material property name, as we don't use it for pressure, saturation, mass fraction etc.

File: modules/porous_flow/src/materials/PorousFlowFluidStateBaseMaterial.C
Comment (cpgr) [reply modules/porous_flow/src/materials/PorousFlowFluidStateBaseMaterial.C:151]: I'll probably change it anyway

File: modules/porous_flow/src/materials/PorousFlowFluidStateSingleComponent.C
Comment (GiudGiud) [inline modules/porous_flow/src/materials/PorousFlowFluidStateSingleComponent.C:160]: could be an idea to propagate that to the other if (the use of constexpr)

File: modules/porous_flow/src/materials/PorousFlowFluidStateSingleComponent.C
Comment (GiudGiud) [inline modules/porous_flow/src/materials/PorousFlowFluidStateSingleComponent.C:164]: better to make those relative to the local pressure (more resilient to regimes / units)
so 1e-5 * _liquid_porepressure[_qp] for example

File: modules/porous_flow/src/materials/PorousFlowFluidStateSingleComponent.C
Comment (GiudGiud) [inline modules/porous_flow/src/materials/PorousFlowFluidStateSingleComponent.C:189]: why 2 dp ? you only added 1 inside the df(p + dp)/dT term

File: modules/porous_flow/src/materials/PorousFlowFluidStateSingleComponent.C
Comment (GiudGiud) [reply modules/porous_flow/src/materials/PorousFlowFluidStateSingleComponent.C:160]: why do you not need second derivatives with AD? AD doesnt do that for you does it?

File: modules/porous_flow/src/materials/PorousFlowFluidState.C
Comment (GiudGiud) [inline modules/porous_flow/src/materials/PorousFlowFluidState.C:109]: is the call right below not setting all the properties?

File: modules/porous_flow/doc/content/source/postprocessors/FVPorousFlowFluidMass.md
Comment (GiudGiud) [inline modules/porous_flow/doc/content/source/postprocessors/FVPorousFlowFluidMass.md:11]:
```suggestion
The fluid component $\kappa$ is specified in the input parameter [!param](/Postprocessors/FVPorousFlowFluidMass/fluid_component).
```

File: modules/porous_flow/doc/content/source/postprocessors/FVPorousFlowFluidMass.md
Comment (GiudGiud) [inline modules/porous_flow/doc/content/source/postprocessors/FVPorousFlowFluidMass.md:14]:
```suggestion
in the [!param](/Postprocessors/FVPorousFlowFluidMass/phase) input parameter.
```

File: modules/porous_flow/doc/content/source/postprocessors/FVPorousFlowFluidMass.md
Comment (GiudGiud) [inline modules/porous_flow/doc/content/source/postprocessors/FVPorousFlowFluidMass.md:17]:
```suggestion
certain saturation, which can be invoked using the [!param](/Postprocessors/FVPorousFlowFluidMass/saturation_threshold) parameter.
```

File: modules/porous_flow/doc/content/source/postprocessors/FVPorousFlowFluidMass.md
Comment (GiudGiud) [inline modules/porous_flow/doc/content/source/postprocessors/FVPorousFlowFluidMass.md:21]:
```suggestion
- For mechanically-coupled simulations, you must provide a [!param](/Postprocessors/FVPorousFlowFluidMass/base_name) parameter that is identical to that used by the TensorMechanics strain calculator, so that this Postprocessor can retrieve the correct strain.  The most common use-case is that you provide no `base_name` to the TensorMechanics strain calculator and hence no `base_name` to this Postprocessor.
```

Comment (GiudGiud) [reply modules/porous_flow/doc/content/source/postprocessors/FVPorousFlowFluidMass.md:21]: once is enough to make sure it doesnt change silently breaking the docs imo

File: modules/porous_flow/include/materials/PorousFlowFluidStateBaseMaterial.h
Comment (GiudGiud) [inline modules/porous_flow/include/materials/PorousFlowFluidStateBaseMaterial.h:48]: so weird that we dont have that elsewhere
@lindsayad do we not?

File: modules/porous_flow/include/materials/PorousFlowFluidStateBaseMaterial.h
Comment (GiudGiud) [reply modules/porous_flow/include/materials/PorousFlowFluidStateBaseMaterial.h:89]: yeah for example!
I think you may want a constexpr in front of is_ad too

File: modules/porous_flow/include/postprocessors/FVPorousFlowFluidMass.h
Comment (GiudGiud) [inline modules/porous_flow/include/postprocessors/FVPorousFlowFluidMass.h:16]:
```suggestion
 * Postprocessor that produces the mass of a given fluid component in a region.
```

File: modules/porous_flow/include/materials/PorousFlowFluidStateBaseMaterial.h
Comment (GiudGiud) [reply modules/porous_flow/include/materials/PorousFlowFluidStateBaseMaterial.h:48]: if we dont, I could see this in MooseTypes.h.
what do you think?

File: modules/porous_flow/src/postprocessors/FVPorousFlowFluidMass.C
Comment (GiudGiud) [inline modules/porous_flow/src/postprocessors/FVPorousFlowFluidMass.C:53]: this logic seems dangerous

if i mispelled base_name, it wont retrieve a total strain and will change the result.

why would not always retrieve the total strain if a base name is passed?

**Review by GiudGiud — COMMENTED** [review]
cursory review

File: modules/porous_flow/include/materials/PorousFlowFluidStateBaseMaterial.h
Comment (lindsayad) [reply modules/porous_flow/include/materials/PorousFlowFluidStateBaseMaterial.h:48]: I would prefer not to. I generally don't like to add more APIs for dropping derivatives. But I would be open to changing my mind

File: modules/porous_flow/include/materials/PorousFlowFluidStateBaseMaterial.h
Comment (cpgr) [reply modules/porous_flow/include/materials/PorousFlowFluidStateBaseMaterial.h:89]: I tried this in the last commit. I couldn't use constexpr though, as it wouldn't compile?

File: modules/porous_flow/src/postprocessors/FVPorousFlowFluidMass.C
Comment (cpgr) [reply modules/porous_flow/src/postprocessors/FVPorousFlowFluidMass.C:53]: Oh, you might want to use a total strain but if you make a typo in the input file, it will just silently not use one and the user won't know why they are getting an incorrect value. I guess we should error if base_name is provided but doesn't actually exist.

This was a copy and paste from the existing post processor, and I think the same logic is used in the energy postprocessor too. I'll make an issue to fix all of these.

File: modules/porous_flow/include/materials/PorousFlowFluidStateBaseMaterial.h
Comment (lindsayad) [reply modules/porous_flow/include/materials/PorousFlowFluidStateBaseMaterial.h:48]: I'm fine with you having it in your module. And if other modules come up with justifiable uses then we could migrate it to the framework

File: modules/porous_flow/src/postprocessors/FVPorousFlowFluidMass.C
Comment (cpgr) [reply modules/porous_flow/src/postprocessors/FVPorousFlowFluidMass.C:53]: I've opened an issue to fix this in the postprocessors not included in this PR

Comment (cpgr) [reply modules/porous_flow/src/postprocessors/FVPorousFlowFluidMass.C:53]: I added a check and a paramError when base_name is given by doesn't exist. I think this is probably the correct behaviour.

File: modules/porous_flow/src/materials/PorousFlowFluidStateSingleComponent.C
Comment (GiudGiud) [inline modules/porous_flow/src/materials/PorousFlowFluidStateSingleComponent.C:155]: should this exclude nodal materials too?

**Review by GiudGiud — COMMENTED** [review]
last set of comments! This looks good.

File: modules/porous_flow/src/materials/PorousFlowFluidStateBaseMaterial.C
Comment (cpgr) [reply modules/porous_flow/src/materials/PorousFlowFluidStateBaseMaterial.C:151]: They are sized in the call to
```
// Size pressure and saturation and prepare the derivative vectors
PorousFlowVariableBaseTempl<is_ad>::computeQpProperties();
```

(I added some inline documentation to make it clear)

File: modules/porous_flow/src/materials/PorousFlowFluidState.C
Comment (cpgr) [reply modules/porous_flow/src/materials/PorousFlowFluidState.C:48]: Yeah, we don't use them in that case

File: modules/porous_flow/src/materials/PorousFlowFluidState.C
Comment (cpgr) [reply modules/porous_flow/src/materials/PorousFlowFluidState.C:242]: Yep. I had to do this earlier though to remove it
```
  if (!_nodal_material)
    if constexpr (!is_ad)
    {
```
I couldn't get these into the one if statement

File: modules/porous_flow/src/materials/PorousFlowFluidStateSingleComponent.C
Comment (cpgr) [reply modules/porous_flow/src/materials/PorousFlowFluidStateSingleComponent.C:155]: Yep, it has the same logic as the comment above

**Review by GiudGiud — APPROVED** [review]

**Comment by GiudGiud** [comment]
looks great!

**Comment by cpgr** [comment]
Thanks for all your help @GiudGiud

**Comment by GiudGiud** [comment]
my pleasure.
You have merge rights now right?

**Comment by cpgr** [comment]
> my pleasure.
> You have merge rights now right?

I think I do, but is it ok to merge my own work as long as someone has reviewed it? And is there a way to mark all of the conversations as resolved?

**Comment by GiudGiud** [comment]
There is no way to mark all conversations as resolved. You have to click them one by one.
Yes it's ok to merge your own work as long as you are the assignee and someone has approved it.
Maybe wait a little on that one, I have to check with Cody on something

File: modules/porous_flow/src/postprocessors/PorousFlowFluidMass.C
Comment (cpgr) [reply modules/porous_flow/src/postprocessors/PorousFlowFluidMass.C:56]: Ended up making a FV-specific one

**Comment by GiudGiud** [comment]
ah yeah forgot I had assigned myself. We dont enforce that one strictly, it's more of a courtesy

**Pull request merged**
